python/dgl/graphbolt/item_sampler.py

```diff
@@ -77,6 +79,70 @@ def minibatcher_default(batch, names):
     return minibatch
 
 
+class ItemShufflerAndBatcher:
+    """A shuffler to shuffle items and create batches.
+
+    This class is used internally by :class:`ItemSampler` to shuffle items and
+    create batches. It is not supposed to be used directly. The intention of
+    this class is to avoid time-consuming iteration over :class:`ItemSet`. As
+    an optimization, it slices from the :class:`ItemSet` via indexing first,
+    then shuffle and create batches.
+
+    Parameters
+    ----------
+    item_set : ItemSet
+        Data to be iterated.
+    shuffle : bool
+        Option to shuffle before batching.
+    batch_size : int
+        The size of each batch.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    buffer_size : int
+        The size of the buffer to store items slices from the :class:`ItemSet`.
+
+    """
+
+    def __init__(
+        self,
+        item_set: ItemSet,
+        shuffle: bool,
+        batch_size: int,
+        drop_last: bool,
+        buffer_size: Optional[bool] = 10 * 1000,
+    ):
+        self._item_set = item_set
+        self._shuffle = shuffle
+        self._batch_size = batch_size
+        self._drop_last = drop_last
+        self._buffer_size = max(buffer_size, 20 * batch_size)
+
+    def __iter__(self):
+        buffer = None
+        num_items = len(self._item_set)
+        curr_idx = 0
+        while curr_idx < num_items:
+            start = curr_idx
+            end = min(curr_idx + self._buffer_size, num_items)
+            buffer = self._item_set[start:end]
+            curr_idx = end
+            indices = torch.arange(end - start)
+            if self._shuffle:
+                np.random.shuffle(indices.numpy())
+            indices_splits = torch.split(indices, self._batch_size)
+            for indices_split in indices_splits:
+                if self._drop_last and len(indices_split) < self._batch_size:
```

Comment:
over drop if _buffer_size & _batch_size != 0.

Reply:
I forgot to round up `buffer_size` to multiply of `batch_size`. If then, no over drop.